test/cpp/tensorexpr/test_registerizer.cpp

```diff
@@ -0,0 +1,543 @@
+#include "test/cpp/tensorexpr/test_base.h"
+
+#include "test/cpp/tensorexpr/test_utils.h"
+#include "torch/csrc/jit/tensorexpr/ir_simplifier.h"
+#include "torch/csrc/jit/tensorexpr/registerizer.h"
+
+#include <iostream>
+
+namespace torch {
+namespace jit {
+using namespace torch::jit::tensorexpr;
+
+// Can replace a simple scalar access with a local variable.
+void testRegisterizerSimple() {
+  KernelScope kernel_scope;
+  Buffer a(BufHandle("A", {1}, kInt));
+  VarHandle x("x", kInt);
+  Stmt* stmt =
+      Block::make({Store::make(a, {0}, 0, 1),
+                   For::make(
+                       x,
+                       0,
+                       10,
+                       Block::make({Store::make(
+                           a, {0}, Add::make(Load::make(a, {0}, 1), x), 1)}))});
+
+  /*
+   * A[0] = 0;
+   * for (int x = 0; x < 10; x++) {
+   *   A[0] = (A[0]) + x;
+   * }
+   */
+
+  registerize(stmt);
+
+  /*
+   * int A_ = 0;
+   * for (int x = 0; x < 10; x++) {
+   *   A_ = x + A_;
+   * }
+   * A[0] = A_;
+   */
+
+  std::ostringstream oss;
+  oss << *stmt;
+
+  const std::string& verification_pattern =
+      R"IR(
+# CHECK: int A_ = 0;
+# CHECK: for (int x = 0; x < 10; x++)
+# CHECK-NOT: A[
+# CHECK:   A_ =
+# CHECK: A[0] = A_;)IR";
+
+  torch::jit::testing::FileCheck().run(verification_pattern, oss.str());
+}
+
+// Won't do replacement of a loop access.
+void testRegisterizerLoop() {
+  KernelScope kernel_scope;
+  Buffer a(BufHandle("A", {10}, kInt));
+  VarHandle x("x", kInt);
+  Stmt* stmt =
+      Block::make({Store::make(a, {0}, 0, 1),
+                   For::make(
+                       x,
+                       0,
+                       10,
+                       Block::make({Store::make(
+                           a, {x}, Add::make(Load::make(a, {x}, 1), x), 1)}))});
+
+  /*
+   * A[0] = 0;
+   * for (int x = 0; x < 10; x++) {
+   *   A[0] = (A[0]) + x;
```

Comment:
Change the comments to reflect the test.

Reply:
Pardon, what do you mean here? How does this comment not reflect it?

Reply:
I thought the code refers to A[x], not A[0]. No?

Reply:
Yes thanks, good catch.